src/SignalR/server/Core/src/Internal/HubMethodDescriptor.cs

```diff
@@ -81,22 +81,53 @@ public HubMethodDescriptor(ObjectMethodExecutor methodExecutor, IServiceProvider
             }
             else if (p.CustomAttributes.Any())
             {
+                var markedParameter = false;
                 foreach (var attribute in p.GetCustomAttributes(true))
                 {
                     if (attribute is IFromServiceMetadata)
                     {
-                        return MarkServiceParameter(index);
+                        ThrowIfMarked(markedParameter);
+                        markedParameter = true;
+                        MarkServiceParameter(index);
                     }
                     else if (attribute is FromKeyedServicesAttribute keyedServicesAttribute)
                     {
-                        if (serviceProviderIsService is IServiceProviderIsKeyedService keyedServiceProvider &&
-                            keyedServiceProvider.IsKeyedService(GetServiceType(p.ParameterType), keyedServicesAttribute.Key))
+                        ThrowIfMarked(markedParameter);
+                        markedParameter = true;
+
+                        if (serviceProviderIsService is IServiceProviderIsKeyedService keyedServiceProvider)
+                        {
+                            if (keyedServiceProvider.IsKeyedService(GetServiceType(p.ParameterType), keyedServicesAttribute.Key))
+                            {
+                                KeyedServiceKeys ??= new List<(int, object)>();
+                                KeyedServiceKeys.Add((index, keyedServicesAttribute.Key));
+                                MarkServiceParameter(index);
+                            }
+                            else
+                            {
+                                throw new InvalidOperationException($"'{p.ParameterType}' is not in DI as a keyed service.");
+                            }
+                        }
+                        else
                         {
-                            KeyedServiceKeys ??= new List<(int, object)>();
-                            KeyedServiceKeys.Add((index, keyedServicesAttribute.Key));
-                            return MarkServiceParameter(index);
+                            throw new InvalidOperationException($"This service provider doesn't support keyed services.");
                         }
                     }
+
+                    void ThrowIfMarked(bool marked)
+                    {
+                        if (marked)
+                        {
+                            throw new InvalidOperationException(
```

Comment:
Maybe we should unify this with the exceptions throw in MVC/minimal APIs?

RE: https://github.com/dotnet/aspnetcore/blob/4a9118c674a798aaf6379b4b7b2d8787bc688f96/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L768

Reply:
That one doesn't mention which method is the culprit, maybe because it can be an unnamed delegate?

I'd like to keep the name for SignalR so it's very clear where the issue is. Maybe just add it at the beginning:
`"{methodExecutor.MethodInfo.FullName}: The {nameof(FromKeyedServicesAttribute)} is not supported on parameters that are also annotated with {nameof(IFromServiceMetadata)}."`

Reply:
Updated the exception message